app/utils/service_client/client.rb

```diff
@@ -0,0 +1,46 @@
+module ServiceClient
+  class Client
+    def initialize(endpoints, middleware = [], http_client = HTTPClient)
+      @_endpoints = endpoints
+      @_context_runner = ContextRunner.new(middleware + [http_client])
+    end
+
+    def get(endpoint, params = {}, opts = {})
+      execute(:get, endpoint, params, opts)
+    end
+
+    def post(endpoint, params = {}, opts = {})
+      execute(:post, endpoint, params, opts)
+    end
+
+    private
+
+    def execute(method, endpoint, params, opts)
+      endpoint_url = @_endpoints[endpoint]
+
+      raise ArgumentError.new("Couldn't find endpoint '#{endpoint}'") if endpoint_url.nil?
+
+      ctx = @_context_runner.execute(
+        url: endpoint_url,
+        method: method,
+        params: params,
+        opts: opts
+      )
```

Comment:
(sorry if I'm commenting too early work to be commented)

There's no reason to lose the logical endpoint name on this level. Down the middleware chain some middleware might like to use it (e.g. validate params based on operation id in swagger schema).